### PR TITLE
Goals Management

### PR DESCRIPTION
**Pull request by aksafan**
## Description

Goals Management

## Jira Related Issue

https://cherise5589.atlassian.net/browse/TASK-28

## Type of Changes

Feature addition

## Acceptance Criteria

All Goals endpoints work

## Update Screenshots
image

## Testing Instructions

Navigate to http://localhost:8000/docs/#/Goals, try all endpoints

### CODE REVIEW

**Review by SonaliBedge — APPROVED** [review]
I’m guessing this hardcoded user ID "7171f91a-bd67-41c2-9e38-7d81be9edf22" is temporary. Later, it will be replaced with all code that’s currently commented out. 
**Note**: It's important to avoid including a large amount of code in a single pull request. Large PRs are harder to review, riskier to merge, and not ideal for collaboration or continuous integration (This isn’t recommended and doesn’t reflect how things are typically done in real-world projects). While it’s okay this time, please keep in mind for the future: try to break your work into smaller, focused PRs—ideally touching just a few files at a time. This makes reviews faster, feedback clearer, and the development process smoother for everyone. 
**Just suggestion**: If possible and time allows in the future, consider adding unit tests for controllers, services, and middleware and to make sure to cover edge cases such as invalid IDs, unauthorized access, and empty payloads.

**Comment by aksafan** [comment]
@SonaliBedge thank you for the comment!

> I’m guessing this hardcoded user ID "7171f91a-bd67-41c2-9e38-7d81be9edf22" is temporary. Later, it will be replaced with all code that’s currently commented out. 

You are 100% right! Getting user from request is being implemented in parallel by @IrinaStoetski, so to not waste time and provide an API for FE team faster, I hard-coded that ID for test purpose for myself and for anyone that will want to test this PR during code review process.
It will be removed in favor of proper code (e.g. [here](https://github.com/Code-the-Dream-School/ii-practicum-team-2-back/pull/10/files#diff-585fd5652f5acca90dea5ea7746c33c2ed1893457f06191c3a0a649fb5745e16R59)).

> **Note**: It's important to avoid including a large amount of code in a single pull request. Large PRs are harder to review, riskier to merge, and not ideal for collaboration or continuous integration (This isn’t recommended and doesn’t reflect how things are typically done in real-world projects). While it’s okay this time, please keep in mind for the future: try to break your work into smaller, focused PRs—ideally touching just a few files at a time. This makes reviews faster, feedback clearer, and the development process smoother for everyone. 

You're absolutely correct and in ideal world it works this way. But during so intensive development process we have right now, it would be a huge overhead to split it so much. Moreover, this PR is the only one resource (goals) management.

> **Just suggestion**: If possible and time allows in the future, consider adding unit tests for controllers, services, and middleware and to make sure to cover edge cases such as invalid IDs, unauthorized access, and empty payloads.

Of course, but not now. There is a plenty of much more important things to do that are really urgent for FE team and project itself. Thus, skipping tests for now is a compelled tradeoff.

File: src/goal/goal.forms.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { z } from "zod";
+
+export const GoalFieldValueForm = z.object({
+  goal_type_field_id: z.string(),
+  value: z.string(),
+});
+export const CreateGoalForm = z.object({
```
Comment (IrinaStoetski) [inline src/goal/goal.forms.ts:7]: Zod uses term "schema" for this https://zod.dev/?id=basic-usage 
Let's be consistent with their naming and name this as CreateGoalFormSchema
And for type - CreateGoalForm
There is no need to add word type in name

File: src/goal/goal.forms.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  name: z.string().min(5).max(256),
+  description: z.string().min(5).max(1024),
```
Comment (IrinaStoetski) [inline src/goal/goal.forms.ts]: 5 symbols for name and description can seem too mush in some cases, let's do 3.

File: src/goal/goal.validators.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  return goal;
+};
+
+export const validateGoalCreationInput = async (input: GoalRequestFormType) => {
+  await validateGoalType(input);
+  await validateGoalFieldValues(input);
+};
+
+async function validateGoalFieldValues(input: GoalRequestFormType) {
```
Comment (IrinaStoetski) [inline src/goal/goal.validators.ts]: `async function validateGoalFieldValues({ goal_type_id }: GoalRequestFormType) {`

File: src/goal/goal.validators.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  const fieldIds = input.goal_field_values.map((f) => f.goal_type_field_id);
+  const matchingFields = await prisma.goalTypeField.findMany({
+    where: {
+      id: { in: fieldIds },
+      goal_type_id: input.goal_type_id,
```
Comment (IrinaStoetski) [inline src/goal/goal.validators.ts]: goal_type_id

Comment (IrinaStoetski) [reply src/goal/goal.validators.ts]: you can do that in other places too

File: src/goal/goal.validators.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  return goal;
+};
+
+export const validateGoalCreationInput = async (input: GoalRequestFormType) => {
+  await validateGoalType(input);
+  await validateGoalFieldValues(input);
+};
+
+async function validateGoalFieldValues(input: GoalRequestFormType) {
+  const fieldIds = input.goal_field_values.map((f) => f.goal_type_field_id);
```
Comment (IrinaStoetski) [inline src/goal/goal.validators.ts]: I would suggest to avoid these 1-letter names for variables. Sky will not fall if we name it as field, but will add a bit of clarity

File: src/goal/goal.validators.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+const validateFieldsMatchGoalType = async (
+  goalTypeId: string,
+  values: UpdateGoalFieldValuesType
+) => {
+  const goalTypeFields = await prisma.goalTypeField.findMany({
+    where: { goal_type_id: goalTypeId },
+  });
+
+  const allowedFieldIds = new Set(goalTypeFields.map((f) => f.id));
```
Comment (IrinaStoetski) [inline src/goal/goal.validators.ts]: the same with 1-letter variables

File: src/errors/domain/domainException.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export default abstract class DomainException extends Error {
```
Comment (IrinaStoetski) [inline src/errors/domain/domainException.ts:13]: would be helpful if there will be small jsdoc description for what cases should we use that

File: src/goal/goal.service.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            })),
+          },
+        },
+        include: {
+          goal_field_values: true,
+        },
+      });
+    } catch (e: unknown) {
+      if (e instanceof PrismaClientKnownRequestError) {
+        if (e.code === "P2002") {
```
Comment (IrinaStoetski) [inline src/goal/goal.service.ts]: should these be somewhere in constants?

File: src/goal/goal.types.ts [hunk 9]
```diff
@@ -1,18 +1,43 @@
...
+export function toGoalResponse(goal: GoalModel): GoalResponse {
+  return {
+    id: goal.id,
+    name: goal.name,
+    description: goal.description,
+    user_id: goal.user_id,
+    goal_type_id: goal.goal_type_id,
+    created_at: goal.created_at.toISOString(),
+    updated_at: goal.updated_at.toISOString(),
+    goal_field_values: goal.goal_field_values.map((m) => ({
```
Comment (IrinaStoetski) [inline src/goal/goal.types.ts]: what is m? let's name it like goalField

**Review by IrinaStoetski — CHANGES_REQUESTED** [review]

File: src/goal/goal.forms.ts (hunk 2 above)
Comment (aksafan) [reply src/goal/goal.forms.ts]: Done!

File: src/goal/goal.validators.ts (hunk 3 above)
Comment (aksafan) [reply src/goal/goal.validators.ts]: Fixed

File: src/goal/goal.validators.ts (hunk 4 above)
Comment (aksafan) [reply src/goal/goal.validators.ts]: Thanks for the hint!
Fixed!

File: src/goal/goal.validators.ts (hunk 6 above)
Comment (aksafan) [reply src/goal/goal.validators.ts]: Done!

File: src/goal/goal.validators.ts (hunk 5 above)
Comment (aksafan) [reply src/goal/goal.validators.ts]: Agreed, thanks! Fixed.

File: src/errors/domain/domainException.ts (hunk 7 above)
Comment (aksafan) [reply src/errors/domain/domainException.ts:13]: Added:

```
/**
 * This is a basic class for all domain exceptions.
 * Extend it to use in code.
 * 
 * Domain exceptions are usually used to communicate between domain and application/presenter layers.
 * In other words, Domain exception - is the way for service to say something to controller.
 * E.g. if there is a business validation error - you can throw ValidationDomainException
 * or NotFoundDomainException if the entity was not found.
 * 
 * Controller can catch and react on these exceptions to show end user the proper response.
 * E.g. send UnprocessableEntity or NotFound response respectively.
 */
 ```

File: src/goal/goal.service.ts (hunk 8 above)
Comment (aksafan) [reply src/goal/goal.service.ts]: Indeed!

Extracted. Thanks!

File: src/goal/goal.types.ts (hunk 9 above)
Comment (aksafan) [reply src/goal/goal.types.ts]: Fixed!

File: src/goal/goal.forms.ts (hunk 1 above)
Comment (aksafan) [reply src/goal/goal.forms.ts:7]: Discussed in PM.

**Review by IrinaStoetski — APPROVED** [review]
